java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,122 @@
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. */
+typealias Identifier = String
+
+/** Lists of [Identifier]s imply an AccessPath. */
+typealias Path = List<Identifier>
+
+/**
+ * [Deduction]s represent how inputs of a Paxel [Expression] contribute to its outputs.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** A representation of a lookup claim in a Paxel [Expression]. */
+  data class Equal(val path: Path = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Return [Equal] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Equal = Equal(
+      path.flatMap { identifier ->
```

Comment:
Let's consider:
```
from x in foo
from y in bar
from z in baz
select x.y.z
```
Because you have a flatmap, I think `x.y.z` will turn into `foo.bar.baz`, which is not correct. While what it should be `foo.y.z`.